docs/current/labs/project-zenith.md

```diff
@@ -0,0 +1,539 @@
+
+```graphql
+{
+  helloWorld {
+    message
+    withName(name: "Monde") {
+      withGreeting(greeting: "Bonjour") {
+        message
+      }
+    }
+  }
+}
+```
+
+The result will be:
+
+```json
+{
+  "helloWorld": {
+    "message": "Hello, World!",
+    "withName": {
+      "withGreeting": {
+        "message": "Bonjour, Monde!"
+      }
+    }
+  }
+}
+```
+
+### Extend core types
+
+You can add a new function to accept and return a `*Container`.
+
+```go
+package main
+
+type Potato struct{}
+
+func (c *Container) AddPotato() *Container {
+  return c.WithNewFile("/potato", ContainerWithNewFileOpts{
+    Contents: "i'm a potato",
+  })
+}
+```
+
+Next, run `dagger mod sync`.
+
+To run the new function, once again use `dagger query` (this example requires a Snyk token):
+
+```sh
+dagger query <<EOF
+{
+  container {
+    from(address:"alpine") {
+      addPotato {
+        withExec(args:["cat", "potato"]) {
+          stdout
+        }
+      }
+    }
+  }
+}
+EOF
+```
+
+## Known issues
+
+* A module's public fields require a `json:"foo"` tag to be queriable.
+* Custom objects in a module require at least one method to be defined on them to be detected by the codegen.
+* When referencing another module as a local dependency, the dependent module must be stored in a sub-directory of the parent module.
+* Custom struct types used as parameters cannot be nested and contain other structs themselves.
+* Calls to functions across modules will be run exactly *once* per-session -- after that, the result will be cached, but only until the next session (a new `dagger query`, etc).
+  * At some point, we will add more fine-grained cache-control.
+
+## Tips and tricks
+
+* The context and error return are optional in the module's function signature; remove them if you don't need them.
+* A module's private fields will not be persisted.
+
+## Troubleshooting
+
+Zenith still isn't perfectly complete! So, if you come across bugs, it helps to have some techniques for working out what's going on.
+
+If you run into problems, please share in the `#zenith-help` channel in the
+[Dagger Discord](https://discord.gg/dagger-io)!
+
+### Rerun commands with `--focus=false`
+
+Sometimes, the dagger client logs are automatically collapsed and don't contain all the information from a failure.
+
+To make sure that logs aren't automatically collapsed, you can run any `dagger` subcommand with the `--focus=false` flag to disable this behavior.
+
+### Access the `docker logs`
+
+The dagger engine runs in a dedicated container. You can find the container using `docker ps`:
+
+```shell
+CONTAINER ID   IMAGE                                                COMMAND                  CREATED       STATUS       PORTS     NAMES
+ffdef3982445   jeremyatdockerhub/dagger-engine-worker-zenith:main   "dagger-entrypoint.s…"   2 hours ago   Up 2 hours             dagger-engine-272478830461fcff
+```
+
+You can then access the logs for the container:
```

Comment:
@jedevc we could also use the approach in https://docs.dagger.io/235290/troubleshooting#dagger-pipeline-is-unresponsive-with-a-buildkit-error to get the container ID.

Reply:
That also works - sorry, I wasn't aware of that section :)